### PR TITLE
RFC: Make avformat decode mjpeg v4l2 with vaapi

### PR DESCRIPTION
**Pull request by viric**
This makes mjpeg decoding on avformat+v4l2 with hwaccel vaapi (common in GNU/Linux specially with Intel CPUs). It works fine here with almost no cpu use on vidloop at 720p 30fps.

All is hardcoded so... how do you see adding some config values to handle this?

Something like:
avformat_hwaccel vaapi           # ffmpeg -hwaccel vaapi ....
avformat_inputformat mjpeg     # ffmpeg -input_format -f v4l2 ...

### CODE REVIEW

**Comment by viric** [comment]
About your comment on using the module swscale for non-hwaccel to decode mjpeg... I need to go from yuvj422p to the dst... But librem doesn't know yuvj422p even to be able to name it.

Maybe we could use swscale internally in avformat to automatically convert to videnc_format, without requiring a filter?

**Comment by alfredh** [comment]
modules/avformat/avformat.c:17:33: fatal error: libavutil/hwcontext.h: No such file or directory
561

perhaps you need to add some compile-time version check.

**Comment by alfredh** [comment]
yuvj422p is yuv422p with JPEG colorspace:

https://github.com/baresip/baresip/blob/master/modules/avcodec/decode.c#L67

**Comment by viric** [comment]
> yuvj422p is yuv422p with JPEG colorspace:

Do you think I could treat yuvj422p as if it was yuv422p to the effect of librem pixfmt conversion, and that would fit?

**Comment by viric** [comment]
> modules/avformat/avformat.c:17:33: fatal error: libavutil/hwcontext.h: No such file or directory
> 561
> 
> perhaps you need to add some compile-time version check.

Would it be fair to disable the hw decoding completely, unless some version is available? Because I don't know the old API to achieve this.

**Comment by alfredh** [comment]
I tested with hwaccel videotoolbox on OSX and it crashes:

```
/vidloop
Enable video-loop on avformat,lavfi,testsrc2: 320 x 240
vidloop: open video display (.)
vidloop: avformat,lavfi,testsrc2: open video source: 320 x 240 at 30.00 fps
Segmentation fault: 11 (core dumped)
```

```
Thread 13 Crashed:
0   libavutil.56.dylib            	0x0000000111d294ce av_hwframe_transfer_data + 94
1   avformat.so                   	0x00000001175f97ac read_thread + 812
2   libsystem_pthread.dylib       	0x00007fffa59e793b _pthread_body + 180
3   libsystem_pthread.dylib       	0x00007fffa59e7887 _pthread_start + 286
4   libsystem_pthread.dylib       	0x00007fffa59e708d thread_start + 13
```

config:

```
avformat_hwaccel        videotoolbox
```

**Comment by viric** [comment]
If you use 'vaapi' but the source is not in an apropriate codec for it to decode, it will crash... Right, I have to handle that!

For example, vaapi and YUY2 from usb webcam, crash. vaapi and MJPG, ok.

**Comment by juha-h** [comment]
viric writes:

> > modules/avformat/avformat.c:17:33: fatal error: libavutil/hwcontext.h: No such file or directory
> > 561
> > 
> > perhaps you need to add some compile-time version check.
> 
> Would it be fair to disable the hw decoding completely, unless some version is available? Because I don't know the old API to achieve this.

When I build baresip avformat module for Android, I do not get that
error.

File: modules/avformat/avformat.c [hunk 1]
```diff
@@ -173,6 +179,21 @@ static int open_codec(struct stream *s, const struct AVStream *strm, int i,
 		return ENOMEM;
 	}
 
+    if (avformat_hwdevice != AV_HWDEVICE_TYPE_NONE)
+    {
```
Comment (alfredh) [inline modules/avformat/avformat.c:183]: please move curly bracket to end of previous line.

also please check the indentation, should be TAB(s)

File: modules/avformat/avformat.c [hunk 2]
```diff
@@ -173,6 +179,21 @@ static int open_codec(struct stream *s, const struct AVStream *strm, int i,
...
+        ret = av_hwdevice_ctx_create(&hwctx, avformat_hwdevice,
+                NULL, NULL, 0);
+        if (ret < 0) {
+            warning("avformat: error opening hw device vaapi (%i)\n", ret);
+            return ENOMEM;
+        }
+
+        ctx->hw_device_ctx = av_buffer_ref(hwctx);
+
+        av_buffer_unref(&hwctx);
```
Comment (alfredh) [inline modules/avformat/avformat.c:194]: perhaps these two lines can be replaced by:

```c
        ctx->hw_device_ctx = hwctx;
```

File: modules/avformat/avformat.c [hunk 3]
```diff
@@ -368,7 +399,19 @@ void avformat_shared_set_video(struct shared *sh, struct vidsrc_st *st)
 
 static int module_init(void)
 {
-	int err;
+    int err;
+    char hwaccel[64] = "";
+
+    conf_get_str(conf_cur(), "avformat_hwaccel", hwaccel, sizeof(hwaccel));
+    if (strnlen(hwaccel, sizeof(hwaccel)) > 0) {
```
Comment (alfredh) [inline modules/avformat/avformat.c:406]: you can also use `str_isset` to check if the string is set, or
check return value of conf_get_str.

**Comment by alfredh** [comment]
regarding the compile-time check for ffmpeg, we have a similar check in avcodec.so:

https://github.com/baresip/baresip/blob/master/modules/avcodec/avcodec.c#L48

**Comment by alfredh** [comment]
hi @viric do you have time to address my review comments ?

**Comment by viric** [comment]
I will I will. I guess there is no hurry.

I'm waiting for a new computer to arrive and engage again refining this.

On Wed, Feb 17, 2021 at 03:38:51AM -0800, Alfred E. Heggestad wrote:
> hi @viric do you have time to address my review comments ?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/baresip/baresip/pull/1216#issuecomment-780498281

-- 
(Escriu-me xifrat si saps PGP / Write ciphered if you know PGP)
PGP key 7CBD1DA5 - https://emailselfdefense.fsf.org/

File: modules/avformat/avformat.c [hunk 4]
```diff
@@ -105,10 +116,10 @@ static void *read_thread(void *data)
 				sys_msleep(1000);
 
 				ret = av_seek_frame(st->ic, -1, 0,
-						    AVSEEK_FLAG_BACKWARD);
+							AVSEEK_FLAG_BACKWARD);
```
Comment (alfredh) [inline modules/avformat/avformat.c:119]: unrelated change

**Comment by alfredh** [comment]
please note that your patch
contains many whitespace changes.

please review your own patch

**Comment by viric** [comment]
oh sorry. I didn't originally care and it seems I used a very inappropriate editor configuration for this project.
Tab characters and 8-column tabs, right?

**Comment by viric** [comment]
Sorry for the inconveniences! I think I fixed the space crippling.
